File: backend/app/schemas/sourcing.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class QuoteCreate(BaseModel):
    vendor: str
    unit_price: float
    oem: str | None = None
    lead_time_weeks: int | None = None
    denominator: str | None = None
    size: float | None = None
    terms_note: str | None = None


class QuoteRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    demand_line_id: uuid.UUID
    vendor: str
    oem: str | None
    unit_price: float
    lead_time_weeks: int | None
    denominator: str | None
    size: float | None
    terms_note: str | None
    state: str
    created_at: datetime


class AwardRequest(BaseModel):
    quote_id: uuid.UUID


class ScopeLineRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    demand_line_id: uuid.UUID
    quote_id: uuid.UUID | None
    vendor: str
    unit_price: float
    qty: int
    change_type: str
    created_at: datetime

File: frontend/src/types/sourcing.ts
export interface CandidateGroup {
  type_query: string
  denominator: string
  size: number
  line_count: number
  total_qty: number
  rom_extended: number | null
  buildings: string[]
  demand_line_ids: string[]
}

export interface CandidatesRead {
  project_id: string
  groups: CandidateGroup[]
  unpoolable_count: number
}

export interface PackageLineRead {
  demand_line_id: string
  qty: number
  target_building: string | null
  target_area: string | null
  state: string
  rom_unit_price: number | null
}

/** One bid, on the same footing as every other bid on the lot. */
export interface LevelingRow {
  quote_id: string
  vendor: string
  oem: string | null
  unit_price: number
  effective_unit: number
  layers: Record<string, number>
  normalized: number
  lead_time_weeks: number | null
  terms_note: string | null
  state: string
  disposition_reason: string | null
  extended: number
  delta_vs_low: number
  delta_vs_low_pct: number | null
  delta_vs_rom: number | null
  delta_vs_rom_pct: number | null
  is_low: boolean
  is_selected: boolean
}

export interface PackageRead {
  id: string
  code: string
  project_id: string
  type_query: string
  denominator: string
  size: number
  state: string
  created_at: string
  line_count: number
  total_qty: number
  rom_unit_price: number | null
  rom_extended: number | null
  quote_count: number
  declined_count: number
  awarded_vendor: string | null
  awarded_extended: number | null
}

export interface PackageDetail {
  package: PackageRead
  lines: PackageLineRead[]
  leveling: LevelingRow[]
}
